internal/pkg/service/appsproxy/proxy/apphandler/authproxy/selector.go

```diff
@@ -0,0 +1,181 @@
+package authproxy
+
+import (
+	"net/http"
+	"net/url"
+	"slices"
+	"strings"
+	"time"
+
+	"github.com/benbjohnson/clock"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/service/appsproxy/config"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/appsproxy/dataapps/auth/provider"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/appsproxy/proxy/pagewriter"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/appsproxy/proxy/requtil"
+	svcErrors "github.com/keboola/keboola-as-code/internal/pkg/service/common/errors"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils/errors"
+)
+
+const (
+	providerCookie     = "_oauth2_provider"
+	providerQueryParam = "provider"
+	callbackQueryParam = "rd"
+	selectionPagePath  = "/_proxy/selection"
+	signOutPath        = "/_proxy/sign_out"
+)
+
+type Selector struct {
+	clock      clock.Clock
+	config     config.Config
+	pageWriter *pagewriter.Writer
+}
+
+func newSelector(d dependencies) *Selector {
+	return &Selector{
+		clock:      d.Clock(),
+		config:     d.Config(),
+		pageWriter: d.PageWriter(),
+	}
+}
+
+// ServeHTTPOrError renders selector page if there is more than one authentication handler,
+// and no handler is selected, or the selected handler is not allowed for the requested path (see api.AuthRule).
+//
+// The selector page is rendered:
+// 1. If it is accessed directly using selectionPagePath, the status code is StatusOK.
+// 2. If no handler is selected and the path requires authorization, the status code is StatusForbidden.
+func (s *Selector) ServeHTTPOrError(handlers map[provider.ID]*Handler, w http.ResponseWriter, req *http.Request) error {
+	// Validate handlers count
+	if len(handlers) == 0 {
+		return svcErrors.NewServiceUnavailableError(errors.New(`no authentication provider found`))
+	}
+
+	// Clear cookie on logout
+	if req.URL.Path == signOutPath {
+		s.clearCookie(w, req)
+	}
+
+	// Shortcut, if there is only one provider
+	if len(handlers) == 1 {
+		for _, handler := range handlers {
+			// Set cookie if needed
+			if providerID := s.providerIDFromCookie(req); providerID != handler.Provider().ID() {
+				s.setCookie(w, req, handler)
+			}
+			return handler.ServeHTTPOrError(w, req)
+		}
+	}
```

Comment:
Is this shortcut very performant? If yes just resolve the comment, however the code will get thiner and therefore there is less possibilities to make mistake, because each handler goes trough same code, here is exception that when single handler provided it does different flow.

Reply:
It is logic shortcut, if there is one handler, no selector page is needed, it is not performance shortcut.

Reply:
I will update comment.